### PR TITLE
Adding app cordon metrics

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward giantswarm/giantswarm#6078

Adding app cordon metrics for alarming. Most of the structure changes introduced in this PR is almost same as in chart-operator.

### CODE REVIEW

File: service/collector/app_resource.go [hunk 1]
```diff
@@ -13,8 +16,8 @@ import (
 )
 
 var (
-	appDesc *prometheus.Desc = prometheus.NewDesc(
-		prometheus.BuildFQName(namespace, "", "app_info"),
+	appDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "app", "info"),
```
Comment (tomahawk28) [inline service/collector/app_resource.go]: Using subsystem as `app` is a better approach I think

File: service/collector/app_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package collector
+
+import (
+	"reflect"
+	"testing"
+	"time"
+)
+
+func Test_convertToTime(t *testing.T) {
```
Comment (tomahawk28) [inline service/collector/app_resource_test.go]: copy same unittests from chart-operator

**Comment by tomahawk28** [comment]
@xh3b4sd I'm looking into aws-operator collector structure, and I see that the difference between collectForAccount and collectAppStatus are having a goroutine or not but not interface as it is. is that right?  https://github.com/giantswarm/aws-operator/blob/027a12a38d51996f9e39ec8d9ded2ed5c0066061/service/collector/trusted_advisor.go#L135

**Comment by xh3b4sd** [comment]
Sorry I did not look closely enough. I see that there is this collector set thingy. This is nice. Let's fix the method order though and then `collectAppStatus` should also have proper error handling. The collector interface supports errors for a reason.

**Comment by tomahawk28** [comment]
@xh3b4sd Thanks, I just got figure it out what was wrong with error handling, PTAL!

File: service/collector/app_resource.go [hunk 3]
```diff
@@ -79,24 +113,38 @@ func (c *AppResource) collectAppStatus(ctx context.Context, ch chan<- prometheus
...
+	if len(split) == 0 {
+		return time.Time{}, microerror.Maskf(invalidExecutionError, "%#q must have at least one item in order to collect metrics for the cordon expiration", datetime)
+	}
 
-	c.logger.LogCtx(ctx, "level", "debug", "message", "finished collecting metrics")
-	return nil
-}
+	t, err := time.Parse(layout, split[0])
+	if err != nil {
+		return time.Time{}, microerror.Maskf(invalidExecutionError, "unavailable to %#q parsing: %#v", split[0], err.Error())
```
Comment (rossf7) [inline service/collector/app_resource.go]:
```suggestion
		return time.Time{}, microerror.Maskf(invalidExecutionError, "parsing timestamp %#q failed: %#v", split[0], err.Error())
```

File: service/collector/app_resource.go [hunk 4]
```diff
@@ -79,24 +113,38 @@ func (c *AppResource) collectAppStatus(ctx context.Context, ch chan<- prometheus
...
+			prometheus.GaugeValue,
+			float64(t.Unix()),
+			key.AppName(app),
+		)
+	}
+	return nil
+}
 
-	c.logger.LogCtx(ctx, "level", "debug", "message", "collecting metrics")
+func convertToTime(datetime string) (time.Time, error) {
```
Comment (rossf7) [inline service/collector/app_resource.go]:
```suggestion
func convertToTime(input string) (time.Time, error) {
```

Nitpick but I find the var name also being a "type" a bit confusing.

File: service/collector/app_resource.go [hunk 5]
```diff
@@ -79,24 +113,38 @@ func (c *AppResource) collectAppStatus(ctx context.Context, ch chan<- prometheus
...
+	layout := "2006-01-02T15:04:05"
```
Comment (rossf7) [inline service/collector/app_resource.go]: In the unit test it has milliseconds `2020-01-02T15:04:05.000`. Is that required? 

We are going to set the annotation in opsctl but I'd like to be clearer on the datetime format.

File: service/collector/app_resource.go [hunk 6]
```diff
@@ -79,24 +113,38 @@ func (c *AppResource) collectAppStatus(ctx context.Context, ch chan<- prometheus
...
 
-	c.collectAppStatus(ctx, ch)
+	split := strings.Split(datetime, ".")
```
Comment (rossf7) [inline service/collector/app_resource.go]: Sorry more questions. Why is `.` the separator here?

File: service/collector/app_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+		{
+			name:     "case 1: normal timestamp parsing",
+			datetime: "2019-12-31T23:59:59.000",
+			expected: expectedTime,
+		},
+		{
+			name:         "case 2: parsing error since unknown ",
+			datetime:     "2019-12-31T23:59:59Z",
+			errorMatcher: IsInvalidExecution,
+		},
```
Comment (rossf7) [inline service/collector/app_resource_test.go]: I think some more failure cases would be good. e.g. 2019-13-31T23:59:59Z

File: service/collector/app_resource.go (hunk 6 above)
Comment (tomahawk28) [reply service/collector/app_resource.go]: This is also from cert-operator that checking the timestamps in secret. AFAIK this is to split millisecond parts from `datetime`. Most of cases we get the timestamps as `2019-12-31T23:59:59.000` but we don't need that part of milliseconds.

Comment (rossf7) [reply service/collector/app_resource.go]: OK cool that makes sense. For the `cordon-until` we don't care about milliseconds so good to strip them. 👍

File: service/collector/app_resource.go (hunk 5 above)
Comment (tomahawk28) [reply service/collector/app_resource.go]: As we discussed below, we would strip milliseconds part.

**Pull request merged**